File: vi/codes/kotlin/chapter_array_and_linkedlist/list.kt
/**
 * File: list.kt
 * Created Time: 2024-01-25
 * Author: curtishd (1023632660@qq.com)
 */

package chapter_array_and_linkedlist

/* Chương trình điều khiển (Driver Code) */
fun main() {
    /* Khởi tạo danh sách (list) */
    // Tập hợp có thể thay đổi (mutable collection)
    val nums = mutableListOf(1, 3, 2, 5, 4)
    println("List nums = $nums")

    /* Truy cập phần tử */
    val num = nums[1]
    println("Access element at index 1, get num = $num")

    /* Cập nhật phần tử */
    nums[1] = 0
    println("Update element at index 1 to 0, get nums = $nums")

    /* Xóa toàn bộ phần tử */
    nums.clear()
    println("After clearing list, nums = $nums")

    /* Thêm phần tử vào cuối danh sách */
    nums.add(1)
    nums.add(3)
    nums.add(2)
    nums.add(5)
    nums.add(4)
    println("After adding elements, nums = $nums")

    /* Chèn phần tử vào giữa danh sách */
    nums.add(3, 6)
    println("Insert number 6 at index 3, get nums = $nums")

    /* Xóa phần tử */
    nums.removeAt(3)
    println("Delete element at index 3, get nums = $nums")

    /* Duyệt danh sách theo chỉ số */
    var count = 0
    for (i in nums.indices) {
        count += nums[i]
    }

    /* Duyệt trực tiếp các phần tử của danh sách */
    for (j in nums) {
        count += j
    }

    /* Nối hai danh sách */
    val nums1 = mutableListOf(6, 8, 7, 10, 9)
    nums.addAll(nums1)
    println("After concatenating list nums1 to nums, get nums = $nums")

    /* Sắp xếp danh sách */
    nums.sort()
    println("After sorting list, nums = $nums")
}

File: vi/codes/typescript/chapter_array_and_linkedlist/list.ts
/**
 * File: list.ts
 * Created Time: 2022-12-10
 * Author: Justin (xiefahit@gmail.com)
 */

/* Khởi tạo danh sách */
const nums: number[] = [1, 3, 2, 5, 4];
console.log(`List nums = ${nums}`);

/* Truy cập phần tử */
const num: number = nums[1];
console.log(`Access element at index 1, get num = ${num}`);

/* Cập nhật phần tử */
nums[1] = 0;
console.log(`Update element at index 1 to 0, get nums = ${nums}`);

/* Xóa toàn bộ danh sách */
nums.length = 0;
console.log(`After clearing list, nums = ${nums}`);

/* Thêm phần tử vào cuối */
nums.push(1);
nums.push(3);
nums.push(2);
nums.push(5);
nums.push(4);
console.log(`After adding elements, nums = ${nums}`);

/* Chèn phần tử vào giữa danh sách */
nums.splice(3, 0, 6);
console.log(`Insert number 6 at index 3, get nums = ${nums}`);

/* Xóa phần tử */
nums.splice(3, 1);
console.log(`Delete element at index 3, get nums = ${nums}`);

/* Duyệt danh sách theo chỉ số */
let count = 0;
for (let i = 0; i < nums.length; i++) {
    count += nums[i];
}
/* Duyệt trực tiếp các phần tử của danh sách */
count = 0;
for (const x of nums) {
    count += x;
}

/* Nối hai danh sách */
const nums1: number[] = [6, 8, 7, 10, 9];
nums.push(...nums1);
console.log(`After concatenating list nums1 to nums, get nums = ${nums}`);

/* Sắp xếp danh sách */
nums.sort((a, b) => a - b);
console.log(`After sorting list, nums = ${nums}`);

export {};
